File: services/storage-writer/src/metrics.rs
use clickhouse::Client;
use domain::{MetricPoint, MetricPointRow, MetricSeries, MetricSeriesRow};

pub async fn insert_metric_series(ch: &Client, series: Vec<MetricSeries>) -> anyhow::Result<()> {
    let mut insert = ch.insert::<MetricSeriesRow>("metric_series").await?;
    for s in series {
        insert.write(&MetricSeriesRow::from(s)).await?;
    }
    insert.end().await?;
    Ok(())
}

pub async fn insert_metric_points(ch: &Client, points: Vec<MetricPoint>) -> anyhow::Result<()> {
    let mut insert = ch.insert::<MetricPointRow>("metric_points").await?;
    for p in points {
        insert.write(&MetricPointRow::from(p)).await?;
    }
    insert.end().await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use domain::{AggregationTemporality, MetricPoint, MetricSeries, MetricType};
    use uuid::Uuid;

    fn make_series(tenant_id: Uuid) -> MetricSeries {
        MetricSeries {
            tenant_id,
            metric_series_id: Uuid::new_v4(),
            metric_name: "http.request.duration".into(),
            description: "Request duration histogram".into(),
            unit: "ms".into(),
            metric_type: MetricType::Histogram,
            is_monotonic: None,
            aggregation_temporality: Some(AggregationTemporality::Cumulative),
            attributes: [("route".to_string(), "/checkout".to_string())]
                .into_iter()
                .collect(),
            resource_attributes: [("host.name".to_string(), serde_json::json!("web-1"))]
                .into_iter()
                .collect(),
            service_name: "checkout".into(),
            environment: "prod".into(),
        }
    }

    fn make_point(tenant_id: Uuid, series_id: Uuid) -> MetricPoint {
        MetricPoint {
            tenant_id,
            metric_series_id: series_id,
            metric_name: "http.request.duration".into(),
            service_name: "checkout".into(),
            time_unix_nano: 1_700_000_000_000_000_000,
            start_time_unix_nano: Some(1_699_000_000_000_000_000),
            value_double: Some(42.5),
            value_int: None,
            histogram_count: Some(10),
            histogram_sum: Some(425.0),
            histogram_bucket_counts: Some(vec![2, 3, 5]),
            histogram_explicit_bounds: Some(vec![10.0, 50.0]),
        }
    }

    #[test]
    fn metric_series_row_maps_type() {
        let series = MetricSeries {
            tenant_id: Uuid::new_v4(),
            metric_series_id: Uuid::new_v4(),
            metric_name: "http.request.duration".into(),
            metric_type: MetricType::Histogram,
            service_name: "checkout".into(),
            ..Default::default()
        };
        let row = MetricSeriesRow::from(series);
        assert_eq!(row.metric_type, "histogram");
    }

    #[test]
    fn metric_series_row_preserves_tenant_id() {
        let tenant_id = Uuid::new_v4();
        let series = MetricSeries {
            tenant_id,
            ..Default::default()
        };
        let row = MetricSeriesRow::from(series);
        assert_eq!(row.tenant_id, tenant_id);
    }

    #[test]
    fn metric_point_row_preserves_tenant_id() {
        let tenant_id = Uuid::new_v4();
        let point = MetricPoint {
            tenant_id,
            ..Default::default()
        };
        let row = MetricPointRow::from(point);
        assert_eq!(row.tenant_id, tenant_id);
    }

    #[test]
    fn metric_series_roundtrip_preserves_all_fields() {
        let tenant_id = Uuid::new_v4();
        let original = make_series(tenant_id);
        let row = MetricSeriesRow::from(original.clone());
        let recovered = MetricSeries::from(row);

        assert_eq!(recovered.tenant_id, original.tenant_id);
        assert_eq!(recovered.metric_series_id, original.metric_series_id);
        assert_eq!(recovered.metric_name, original.metric_name);
        assert_eq!(recovered.description, original.description);
        assert_eq!(recovered.unit, original.unit);
        assert_eq!(recovered.metric_type, original.metric_type);
        assert_eq!(recovered.is_monotonic, original.is_monotonic);
        assert_eq!(
            recovered.aggregation_temporality,
            original.aggregation_temporality
        );
        assert_eq!(recovered.attributes, original.attributes);
        assert_eq!(recovered.resource_attributes, original.resource_attributes);
        assert_eq!(recovered.service_name, original.service_name);
        assert_eq!(recovered.environment, original.environment);
    }

    #[test]
    fn metric_series_type_roundtrips_all_variants() {
        for (metric_type, expected_str) in [
            (MetricType::Gauge, "gauge"),
            (MetricType::Sum, "sum"),
            (MetricType::Histogram, "histogram"),
            (MetricType::ExponentialHistogram, "exponential_histogram"),
            (MetricType::Summary, "summary"),
        ] {
            let row = MetricSeriesRow::from(MetricSeries {
                metric_type: metric_type.clone(),
                ..Default::default()
            });
            assert_eq!(row.metric_type, expected_str);
            let recovered = MetricSeries::from(row);
            assert_eq!(recovered.metric_type, metric_type);
        }
    }

    #[test]
    fn metric_point_roundtrip_preserves_all_fields() {
        let tenant_id = Uuid::new_v4();
        let series_id = Uuid::new_v4();
        let original = make_point(tenant_id, series_id);
        let row = MetricPointRow::from(original.clone());
        let recovered = MetricPoint::from(row);

        assert_eq!(recovered.tenant_id, original.tenant_id);
        assert_eq!(recovered.metric_series_id, original.metric_series_id);
        assert_eq!(recovered.metric_name, original.metric_name);
        assert_eq!(recovered.service_name, original.service_name);
        assert_eq!(recovered.time_unix_nano, original.time_unix_nano);
        assert_eq!(
            recovered.start_time_unix_nano,
            original.start_time_unix_nano
        );
        assert_eq!(recovered.value_double, original.value_double);
        assert_eq!(recovered.value_int, original.value_int);
        assert_eq!(recovered.histogram_count, original.histogram_count);
        assert_eq!(recovered.histogram_sum, original.histogram_sum);
        assert_eq!(
            recovered.histogram_bucket_counts,
            original.histogram_bucket_counts
        );
        assert_eq!(
            recovered.histogram_explicit_bounds,
            original.histogram_explicit_bounds
        );
    }
}

File: scripts/seed/metrics.py
from __future__ import annotations

import json
import random
import uuid
from datetime import datetime, timezone

from seed.world import Service, Tenant

SERIES_COLS = [
    "tenant_id", "metric_series_id", "metric_name", "description", "unit",
    "metric_type", "is_monotonic", "aggregation_temporality",
    "attributes", "resource_attributes", "service_name", "environment", "created_at",
]

POINT_COLS = [
    "tenant_id", "metric_series_id", "metric_name", "service_name",
    "time_unix_nano", "start_time_unix_nano",
    "value_double", "value_int",
    "histogram_count", "histogram_sum",
    "histogram_bucket_counts", "histogram_explicit_bounds",
]

# Fixed metric definitions emitted by every service
_METRIC_DEFS = [
    ("http.server.request_count", "Total HTTP requests", "requests", "sum", 1, "DELTA"),
    ("http.server.error_count",   "Total HTTP errors",   "requests", "sum", 1, "DELTA"),
    ("http.server.duration",      "Request duration",    "ms",       "histogram", None, "DELTA"),
    ("process.memory.usage",      "Memory usage",        "bytes",    "gauge",     None, None),
]


def make_series_rows(tenant: Tenant, service: Service) -> list[list]:
    """Return one metric_series row per metric definition for this service.

    Series IDs are deterministic from tenant_id + service_name + metric_name
    so repeated calls produce the same UUIDs (idempotent for ReplacingMergeTree).
    """
    rows = []
    now_str = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")
    for name, desc, unit, mtype, is_mono, temporality in _METRIC_DEFS:
        series_id = str(uuid.uuid5(
            uuid.NAMESPACE_OID,
            f"{tenant.tenant_id}:{service.service_name}:{name}",
        ))
        rows.append([
            tenant.tenant_id,
            series_id,
            name,
            desc,
            unit,
            mtype,
            is_mono,
            temporality,
            json.dumps({}),
            json.dumps({"service.name": service.service_name}),
            service.service_name,
            tenant.environment,
            now_str,
        ])
    return rows


def generate_metric_points(
    tenant: Tenant,
    service: Service,
    series_rows: list[list],
    dt: datetime,
    request_count: int,
    rng: random.Random,
) -> list[list]:
    """Return one metric_point row per series for the given 15-second tick."""
    ts_ns = int(dt.timestamp() * 1e9)
    error_count = max(0, int(request_count * rng.uniform(0.005, 0.05)))
    mem_bytes = rng.randint(64 * 1024 * 1024, 512 * 1024 * 1024)

    series_map = {r[SERIES_COLS.index("metric_name")]: r for r in series_rows}
    rows = []

    def _sid(name: str) -> str:
        return series_map[name][SERIES_COLS.index("metric_series_id")]

    # http.server.request_count — counter (value_int)
    rows.append([
        tenant.tenant_id, _sid("http.server.request_count"), "http.server.request_count", service.service_name,
        ts_ns, ts_ns, None, request_count, None, None, [], [],
    ])

    # http.server.error_count — counter (value_int)
    rows.append([
        tenant.tenant_id, _sid("http.server.error_count"), "http.server.error_count", service.service_name,
        ts_ns, ts_ns, None, error_count, None, None, [], [],
    ])

    # http.server.duration — histogram
    bounds = [5.0, 10.0, 25.0, 50.0, 100.0, 250.0, 500.0, 1000.0, 2500.0, 5000.0]
    bucket_counts = [rng.randint(0, max(1, request_count // 10)) for _ in range(len(bounds) + 1)]
    total_dur = sum(b * c for b, c in zip(bounds, bucket_counts[:-1]))
    rows.append([
        tenant.tenant_id, _sid("http.server.duration"), "http.server.duration", service.service_name,
        ts_ns, ts_ns, None, None,
        sum(bucket_counts), total_dur,
        bucket_counts, bounds,
    ])

    # process.memory.usage — gauge (value_int)
    rows.append([
        tenant.tenant_id, _sid("process.memory.usage"), "process.memory.usage", service.service_name,
        ts_ns, None, None, mem_bytes, None, None, [], [],
    ])

    return rows
